### ISSUE TITLE
Provide an API for overriding scope tracking

### ISSUE THREAD

**Issue by Alxandr** [tokio-rs/tracing#426]
## Feature Request
The new Registry type enables splitting different responsibilities between different implementors with "layers" wrapped around it. However, if you need to swap out the way the current span is tracked (with enter/exit). The only way to do so would be to create your own "Registry" type and enable wrapping the same layers around it.

Instead, I suggest creating a trait for this functionality, extracted from the registry:

```rust
pub trait Scope {
  fn enter(&self, span: Id);
  fn exit(&self, span: Id);
  fn current_span(&self) -> Option<Id>;
}
```

The registry can then carry a `Scope` impl, and just delegate to it.

### Crates

* `tracing-subscriber`

### Motivation

Enabling things like providing a task-local backed `Scope` impl.

### Alternatives

The alternatives are generally either fairly ugly (see [example](https://gist.github.com/Alxandr/061897f337f3977dac0b3ae88f2622aa#file-tracing-std-async-experiment-rs-L96-L98)), or requires copying a bunch of functionality in from `tracing-subscriber`.

**Comment by hawkw** [comment]
I'm definitely on board with this — it would also help significantly with `no_std` support in the `tracing-subscriber` crate, since this would allow swapping out the use of thread-locals for something else, such as CPU ids in a bare-metal OS use case, or just having a single context in an embedded use case with only one core, etc.

> ### Motivation
> 
> Enabling things like providing a task-local backed `Scope` impl.

Just to double check, you determined that using `async-std`'s task locals didn't really have any advantages over the current thread-local approach, right? If I recall correctly, the granularity of a _task_-local (rather than a "future-local") still resulted in surprising/wrong scopes for async fns when not using `instrument`?

**Comment by Alxandr** [comment]
I've not determined that. I've determined that using task-locals wasn't enough to get rid of task instrumenting. I still think using task-local is better than thread-local, cause a single task could (as far as I know) move between threads. It might be an unwarranted fear in my case though, but comming from .NET, there's a lot of issues with `ThreadLocal<>` that `AsyncLocal<>` solves. The situation in rust is slightly different though, given that you always have to poll on the root future...

**Comment by hawkw** [comment]
@Alxandr 
> The situation in rust is slightly different though, given that you always have to poll on the root future...

yup, and I think that neatly solves the problem you described with thread-locals. :)
